### PR TITLE
Node/ghc 9.2.4

### PR DESCRIPTION
**Pull request by newhoggy**

### CODE REVIEW

File: bench/tx-generator/src/Cardano/Benchmarking/Script/Core.hs [hunk 1]
```diff
@@ -142,7 +141,7 @@ getConnectClient = do
   tracers  <- get BenchTracers
   (Testnet networkMagic) <- getUser TNetworkId
   protocol <- get Protocol
-  void $ return $(btSubmission2_ tracers)
+  void $ return $ btSubmission2_ tracers -- TODO this line looks strange
```
Comment (angerman) [inline bench/tx-generator/src/Cardano/Benchmarking/Script/Core.hs:144]: @newhoggy Isn't this equivalent to
```
 _ <- pure $ btSubmission2_ tracers
 ```
 Yes, that indeed looks odd.

File: cardano-api/src/Cardano/Api/Fees.hs [hunk 2]
```diff
@@ -1098,8 +1098,12 @@ makeTransactionBodyAutoBalance eraInMode systemstart history pparams
                 returnCollateral = fromShelleyLovelace . Ledger.rationalToCoinViaFloor $ amt % 100
 
             case (txReturnCollateral, txTotalCollateral) of
+#if MIN_VERSION_base(4,16,0)
+#else
+              -- TODO figure out how to include the following cases in ghc-9.2.4
               (rc@TxReturnCollateral{}, tc@TxTotalCollateral{}) ->
                 (rc, tc)
+#endif
```
Comment (angerman) [inline cardano-api/src/Cardano/Api/Fees.hs:1106]: We should log the error message here.
```
src/Cardano/Api/Fees.hs:1104:15: error: [-Woverlapping-patterns, -Werror=overlapping-patterns]
    Pattern match is redundant
    In a case alternative:
        (rc@TxReturnCollateral {}, tc@TxTotalCollateral {}) -> ...
     |
1104 |               (rc@TxReturnCollateral{}, tc@TxTotalCollateral{}) ->
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

And yes, that one _is_ redundant. It's the same as

https://github.com/input-output-hk/cardano-node/pull/4472/files#diff-12e0bb9f3db22ac71c93ed3932f54c91a6f6bf7243afaa29ccd7ecb3657e6069L1074
```
   calcReturnAndTotalCollateral _ _ _ rc@TxReturnCollateral{} tc@TxTotalCollateral{} _ _ = (rc,tc)
```
which will fire before we even end in this case.

File: cardano-api/src/Cardano/Api/InMode.hs [hunk 3]
```diff
@@ -155,6 +155,8 @@ toConsensusGenTx (TxInMode (ShelleyTx _ tx) BabbageEraInCardanoMode) =
   where
     tx' = Consensus.mkShelleyTx tx
 
+toConsensusGenTx _ = error "TODO: toConsensusGenTx"
+
```
Comment (angerman) [inline cardano-api/src/Cardano/Api/InMode.hs:159]:
```
src/Cardano/Api/InMode.hs:110:1: error: [-Wincomplete-patterns, -Werror=incomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘toConsensusGenTx’:
        Patterns of type ‘TxInMode mode’ not matched:
            TxInMode (ShelleyTx _ _) ByronEraInByronMode
            TxInMode (ShelleyTx _ _) ByronEraInCardanoMode
    |
110 | toConsensusGenTx (TxInMode (ByronTx tx) ByronEraInByronMode) =
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
 ```

File: cardano-api/src/Cardano/Api/Tx.hs [hunk 4]
```diff
@@ -143,6 +143,8 @@ instance Eq (Tx era) where
 
     (==) ByronTx{} (ShelleyTx era _) = case era of {}
 
+    (==) _ _ = error "Error: Eq (Tx era)"
+
```
Comment (angerman) [inline cardano-api/src/Cardano/Api/Tx.hs:147]:
```
src/Cardano/Api/Tx.hs:132:5: error: [-Wincomplete-patterns, -Werror=incomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘==’:
        Patterns of type ‘Tx era’, ‘Tx era’ not matched:
            (ShelleyTx _ _) (ByronTx _)
    |
132 |     (==) (ByronTx txA)
    |     ^^^^^^^^^^^^^^^^^^...
```

File: cardano-cli/src/Cardano/CLI/Byron/Legacy.hs [hunk 5]
```diff
@@ -4,6 +4,8 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
+{-# OPTIONS_GHC -Wno-dodgy-imports #-}
+
```
Comment (angerman) [inline cardano-cli/src/Cardano/CLI/Byron/Legacy.hs:8]: Shouldn't this better be in `cabal.project`?

File: cardano-cli/src/Cardano/CLI/Byron/Parsers.hs [hunk 6]
```diff
@@ -21,7 +21,7 @@ module Cardano.CLI.Byron.Parsers
   , parseUpdateVoteThd
   ) where
 
-import           Cardano.Prelude hiding (option)
+import           Cardano.Prelude
```
Comment (angerman) [inline cardano-cli/src/Cardano/CLI/Byron/Parsers.hs:24]: This change seems to force the `option` -> `Opt.option` changes below?

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 7]
```diff
@@ -3,6 +3,8 @@
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
 
+{-# OPTIONS_GHC -Wno-deprecations #-}
+
```
Comment (angerman) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:7]: Again, shouldn't this be in the `cabal.project`?

File: cardano-node/src/Cardano/Node/Run.hs [hunk 8]
```diff
@@ -13,6 +13,8 @@
 #define UNIX
 #endif
 
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
```
Comment (angerman) [inline cardano-node/src/Cardano/Node/Run.hs:17]: `cabal.project`?

File: cardano-node/src/Cardano/Node/Tracing/Tracers/ChainDB.hs [hunk 9]
```diff
@@ -7,6 +7,7 @@
 {-# LANGUAGE UndecidableInstances #-}
 
 {-# OPTIONS_GHC -Wno-orphans  #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
```
Comment (angerman) [inline cardano-node/src/Cardano/Node/Tracing/Tracers/ChainDB.hs:10]: `cabal.project`, maybe I'm wrong?

File: cardano-node/src/Cardano/Node/Tracing/Tracers/ChainDB.hs [hunk 10]
```diff
@@ -55,6 +56,7 @@ import qualified Data.Aeson.KeyMap as KeyMap
 import qualified Ouroboros.Network.AnchoredFragment as AF
 
 {-# ANN module ("HLint: ignore Redundant bracket" :: Text) #-}
```
Comment (angerman) [inline cardano-node/src/Cardano/Node/Tracing/Tracers/ChainDB.hs:58]: this _must_ be a `HLINT` annotation. `ANN` will *force* TH, though it's kinda orthogonal to this PR. Just came across it.

**Review by angerman — CHANGES_REQUESTED** [review]
This is great @newhoggy!

File: cardano-cli/src/Cardano/CLI/Byron/Parsers.hs (hunk 6 above)
Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/Byron/Parsers.hs:24]: For some reason I had to do this for `ghc-8.10.7`.

File: cardano-node/src/Cardano/Node/Tracing/Tracers/ChainDB.hs (hunk 10 above)
Comment (newhoggy) [reply cardano-node/src/Cardano/Node/Tracing/Tracers/ChainDB.hs:58]: Yeah, we should replace those with the new `HLINT` annotations.

**Pull request closed without merging**